src/function/cast_rules.cpp

```diff
@@ -207,6 +207,15 @@ int64_t CastRules::ImplicitCast(const LogicalType &from, const LogicalType &to)
 		// if aliases are different, an implicit cast is not possible
 		return -1;
 	}
+	if (from.id() == LogicalTypeId::LIST && to.id() == LogicalTypeId::LIST) {
+		// Lists can be cast if their child types can be cast
+		auto child_cost = ImplicitCast(ListType::GetChildType(from), ListType::GetChildType(to));
+		if (child_cost >= 100) {
+			// subtract one from the cost because we prefer LIST[X] -> LIST[VARCHAR] over LIST[X] -> VARCHAR
+			child_cost--;
+		}
+		return child_cost;
+	}
 	if (from.id() == to.id()) {
```

Comment:
I spotted something weird here a couple weeks ago, since this takes care of `id == id`, then the if statement you removed was dead code. But below that, there is another similar case like this for UNION - which is also dead code right?

Reply:
Yes it is. I tried to move it up but it did not work (as dead code often doesn't!) so I left it there for now. It should indeed be removed or reworked.